// react-laravel-paginex/dist/Pagination.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Pagination = /*#__PURE__*/function (_Component) {
  _inherits(Pagination, _Component);

  var _super = _createSuper(Pagination);

  function Pagination(_props) {
    var _this;

    _classCallCheck(this, Pagination);

    _this = _super.call(this, _props);

    _defineProperty(_assertThisInitialized(_this), "getProps", function (props) {
      var defaultProps = Pagination.defaultProps.options;
      var options = _this.props.options;
      Object.keys(defaultProps).forEach(function (key) {
        options[key] = props[key] ? props[key] : props['options'][key] ? props['options'][key] : defaultProps[key];
      });

      _this.setState({
        options: options,
        paginationData: props.data
      });
    });

    _defineProperty(_assertThisInitialized(_this), "isCurrent", function (page) {
      var currentPage = _this.state.paginationData.meta ? _this.state.paginationData.meta.current_page : _this.state.paginationData.current_page;
      return currentPage === page;
    });

    _defineProperty(_assertThisInitialized(_this), "handleClick", function (page) {
      var parameters = {};

      if (_this.props.requestParams) {
        parameters = _this.props.requestParams;
      }

      parameters.page = page;

      _this.props.changePage(parameters);
    });

    _defineProperty(_assertThisInitialized(_this), "generateButtonsPrev", function () {
      var options = _this.state.options;

      if (options.buttonIcons) {
        return /*#__PURE__*/_react["default"].createElement("i", {
          className: options.prevButtonIcon
        });
      }

      return options.prevButtonText;
    });

    _defineProperty(_assertThisInitialized(_this), "generateButtonsNext", function () {
      var options = _this.state.options;

      if (options.buttonIcons) {
        return /*#__PURE__*/_react["default"].createElement("i", {
          className: options.nextButtonIcon
        });
      }

      return options.nextButtonText;
    });

    _defineProperty(_assertThisInitialized(_this), "generatePagination", function () {
      var paginationData = _this.state.paginationData;
      var pagination;

      if (Object.keys(paginationData).length) {
        var options = _this.state.options;
        var current = paginationData.hasOwnProperty('current_page') ? paginationData.current_page : paginationData.meta.current_page,
            last = paginationData.hasOwnProperty('last_page') ? paginationData.last_page : paginationData.meta.last_page,
            delta = parseInt(options.numbersCountForShow),
            left = current - delta,
            right = current + delta + 1,
            range = [],
            rangeWithDots = [],
            l;

        for (var i = 1; i <= last; i++) {
          if (i === 1 || i === last || i >= left && i < right) {
            range.push(i);
          }
        }

        for (var _i = 0, _range = range; _i < _range.length; _i++) {
          var _i2 = _range[_i];

          if (l) {
            if (_i2 - l === 2) {
              rangeWithDots.push(l + 1);
            } else if (_i2 - l !== 1) {
              rangeWithDots.push('...');
            }
          }

          rangeWithDots.push(_i2);
          l = _i2;
        }

        var nextPageUrl = paginationData.hasOwnProperty('next_page_url') ? paginationData.next_page_url : paginationData.links.next;
        var prevPageUrl = paginationData.hasOwnProperty('prev_page_url') ? paginationData.prev_page_url : paginationData.links.prev;
        pagination = /*#__PURE__*/_react["default"].createElement("ul", {
          className: options.containerClass
        }, prevPageUrl ? /*#__PURE__*/_react["default"].createElement("li", {
          className: options.prevButtonClass,
          onClick: function onClick(event) {
            event.preventDefault();

            _this.handleClick(current - 1);
          }
        }, /*#__PURE__*/_react["default"].createElement("a", {
          href: "",
          className: options.numberClass
        }, _this.generateButtonsPrev())) : "", rangeWithDots.map(function (page, index) {
          return _this.generateNumber(page, index);
        }), nextPageUrl ? /*#__PURE__*/_react["default"].createElement("li", {
          className: options.nextButtonClass,
          onClick: function onClick(event) {
            event.preventDefault();

            _this.handleClick(current + 1);
          }
        }, /*#__PURE__*/_react["default"].createElement("a", {
          href: "",
          className: options.numberClass
        }, _this.generateButtonsNext())) : "");
      }

      return pagination;
    });

    _this.state = {
      options: {},
      paginationData: {},
      nextPageUrl: null,
      prevPageUrl: null,
      currentPage: null
    };
    return _this;
  }

  _createClass(Pagination, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.getProps(this.props);
    }
  }, {
    key: "UNSAFE_componentWillReceiveProps",
    value: function UNSAFE_componentWillReceiveProps(props, nextContext) {
      this.getProps(props);
    } // Transform props

  }, {
    key: "generateNumber",
    value: function generateNumber(page, index) {
      var _this2 = this;

      var options = this.state.options;
      return /*#__PURE__*/_react["default"].createElement("li", {
        className: this.isCurrent(page) ? options.numberButtonClass + " " + options.activeClass : options.numberButtonClass,
        key: index
      }, /*#__PURE__*/_react["default"].createElement("a", {
        href: "",
        className: options.numberClass,
        onClick: function onClick(event) {
          event.preventDefault();

          _this2.handleClick(page === '...' ? index + 1 : page);
        }
      }, page));
    }
  }, {
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, this.generatePagination());
    }
  }]);

  return Pagination;
}(_react.Component);

Pagination.defaultProps = {
  options: {
    containerClass: "pagination",
    buttonIcons: false,
    prevButtonClass: "page-item",
    prevButtonText: "Prev",
    prevButtonIcon: "fa fa-chevron-left",
    nextButtonClass: "page-item",
    nextButtonText: "Next",
    nextButtonIcon: "fa fa-chevron-right",
    numberButtonClass: "page-item",
    numberClass: "page-link",
    numbersCountForShow: 2,
    activeClass: 'active'
  },
  data: {}
};
Pagination.propTypes = {
  options: _propTypes["default"].shape({
    containerClass: _propTypes["default"].string,
    buttonIcons: _propTypes["default"].bool,
    nextButtonClass: _propTypes["default"].string,
    nextButtonText: _propTypes["default"].string,
    nextButtonIcon: _propTypes["default"].string,
    prevButtonClass: _propTypes["default"].string,
    prevButtonText: _propTypes["default"].string,
    prevButtonIcon: _propTypes["default"].string,
    numberButtonClass: _propTypes["default"].string,
    numberClass: _propTypes["default"].string,
    numberCountForShow: _propTypes["default"].number,
    activeClass: _propTypes["default"].string
  }),
  data: _propTypes["default"].object
};
var _default = Pagination;
exports["default"] = _default;
